### PR TITLE
refactor(protocol-engine): rework command models for HTTP API schema

### PR DESCRIPTION
**Pull request by mcous**
## Overview

This PR is precursor work to #7870 and #7871.

Since the Protocol Engine is the primary processor and state-tracker of protocol commands, we want the models that it uses to be broadly compatible with the HTTP API. This PR reworks all the Protocol Engine's commands to (mostly) match the Command model from existing sessions.

This is a massive PR because it turns out, there was a lot of not-single-responsibility stuff about commands in the Protocol Engine. This PR, by necessity, attempts to sequester some of those responsibilities to prevent this sort of hellish churn in the future.

## Changelog

- refactor(protocol-engine): rework command models for HTTP API schema

## Review requests

This is awful and I'm sorry. When in doubt, fire-drill me in Slack for an in-person call to resolve stuff.

### Important smoke tests

I've run the following sessions to completion on my robot on commit b011d29:

- [x] Test that all the sessions other than `liveProtocol` are working with the `enableProtocolEngine` feature flag **disabled**
    - [x] `calibrationCheck`
    - [x] `tipLengthCalibration`
    - [x] `deckCalibration`
    - [x] `pipetteOffsetCalibration`
    - [x] `protocol`
- [x] Test that JSON protocols are still able to run on the engine via HTTP with the `enableProtocolEngine` feature flag **enabled**
    - Fundamentally works, with the caveats below

### Unrelated issues found during testing

Not going to address any of these in this PR, but noting them to make sure we don't lose them:

- ~Protocol engine does not support drop tip in fixed trash from JSON protocol~
    - Fix: check for `fixedTrash` quirk rather than looking for the hardcoded `FIXED_TRASH_ID`
    - Accidentally included this fix in this PR. I regret nothing
- Protocol engine does not home pipettes before usage
    - Fix: not sure, but seems like a hardware controller job?
    - Figure out if the engine is using the HC incorrectly  

## Risk assessment

**Medium**. This PR gets close to existing sessions endpoints by necessity. **It (intentionally) breaks the non-production `liveProtocol` sessions** in favor of the upcoming feature-flagged `basic` session. Also note that `liveProtocol` sessions **are not the same thing as the `protocol` sessions** documented in the [http-api-beta][]. These protocol-file based sessions should continue to function.

Mitigation steps:

- Integration tests for all production sessions are still passing
- I've made sure that I can still run every session on my robot while running this branch

[http-api-beta]: https://github.com/Opentrons/http-api-beta

### CODE REVIEW

File: api/src/opentrons/protocol_engine/state/geometry.py [hunk 1]
```diff
@@ -173,8 +172,13 @@ def get_tip_drop_location(
         pipette_config: PipetteDict,
     ) -> WellLocation:
         """Get tip drop location given labware and hardware pipette."""
-        # return to top if labware is fixed trash
-        if labware_id == FIXED_TRASH_ID:
+        # return to top if labware is fixed trash=
+        is_fixed_trash = self._labware.get_labware_has_quirk(
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/state/geometry.py:176]: Oh oops, this is the fix I said I wasn't going to include. Accidentally committed; I think I wanna leave it in because I'm feeling lazy and it's a very small fix

File: api/src/opentrons/protocol_engine/commands/add_labware_definition.py [hunk 2]
```diff
@@ -1,22 +1,19 @@
...
+
 from opentrons.protocols.models import LabwareDefinition
+from .command import AbstractCommandImpl, BaseCommand, BaseCommandRequest
+
+AddLabwareDefinitionCommandType = Literal["addLabwareDefinition"]
 
 
-class AddLabwareDefinitionRequest(BaseModel):
+class AddLabwareDefinitionData(BaseModel):
     """Request to add a labware definition."""
```
Comment (sanni-t) [inline api/src/opentrons/protocol_engine/commands/add_labware_definition.py]:
```suggestion
    """Data required to add a labware definition."""
```

File: api/src/opentrons/protocol_engine/commands/aspirate.py [hunk 3]
```diff
@@ -1,41 +1,57 @@
...
+from .pipetting_common import BaseLiquidHandlingData, BaseLiquidHandlingResult
+from .command import AbstractCommandImpl, BaseCommand, BaseCommandRequest
 
-@final
-class AspirateRequest(BaseLiquidHandlingRequest):
+AspirateCommandType = Literal["aspirate"]
+
+
+class AspirateData(BaseLiquidHandlingData):
     """A request to move to a specific well and aspirate from it."""
```
Comment (sanni-t) [inline api/src/opentrons/protocol_engine/commands/aspirate.py]:
```suggestion
    """Data required to aspirate from a specific well."""
```

File: api/src/opentrons/protocol_engine/commands/dispense.py [hunk 4]
```diff
@@ -1,41 +1,57 @@
...
+from .command import AbstractCommandImpl, BaseCommand, BaseCommandRequest
 
 
-@final
-class DispenseRequest(BaseLiquidHandlingRequest):
+DispenseCommandType = Literal["dispense"]
+
+
+class DispenseData(BaseLiquidHandlingData):
     """A request to aspirate from a specific well."""
```
Comment (sanni-t) [inline api/src/opentrons/protocol_engine/commands/dispense.py]:
```suggestion
    """Data required to dispense to a specific well."""
```

File: api/src/opentrons/protocol_engine/commands/drop_tip.py [hunk 5]
```diff
@@ -1,17 +1,19 @@
...
-from .pipetting_common import BasePipettingRequest
+from .pipetting_common import BasePipettingData
+from .command import AbstractCommandImpl, BaseCommand, BaseCommandRequest
 
+DropTipCommandType = Literal["dropTip"]
 
-class DropTipRequest(BasePipettingRequest):
+
+class DropTipData(BasePipettingData):
     """A request to drop a tip in a specific well."""
```
Comment (sanni-t) [inline api/src/opentrons/protocol_engine/commands/drop_tip.py]:
```suggestion
    """Data required to drop a tip in a specific well."""
```

File: api/src/opentrons/protocol_engine/commands/load_labware.py [hunk 6]
```diff
@@ -1,16 +1,18 @@
...
-from .command import CommandImplementation, CommandHandlers
+from opentrons.protocol_engine.types import LabwareLocation
+from .command import AbstractCommandImpl, BaseCommand, BaseCommandRequest
+
+LoadLabwareCommandType = Literal["loadLabware"]
 
 
-class LoadLabwareRequest(BaseModel):
+class LoadLabwareData(BaseModel):
     """A request to load a labware into a slot."""
```
Comment (sanni-t) [inline api/src/opentrons/protocol_engine/commands/load_labware.py]:
```suggestion
    """Data required to load a labware into a slot."""
```

File: api/src/opentrons/protocol_engine/commands/pipetting_common.py [hunk 7]
```diff
@@ -4,7 +4,7 @@
 from ..types import WellLocation
 
 
-class BasePipettingRequest(BaseModel):
+class BasePipettingData(BaseModel):
     """Base class for pipetting requests that interact with wells."""
```
Comment (sanni-t) [inline api/src/opentrons/protocol_engine/commands/pipetting_common.py]:
```suggestion
    """Base class for data payloads of commands that interact with wells."""
```

**Review by sanni-t — COMMENTED** [review]
Still making my way through the PR but posting these so I don't lose them

File: api/src/opentrons/protocol_engine/commands/pipetting_common.py [hunk 8]
```diff
@@ -21,13 +21,13 @@ class BasePipettingRequest(BaseModel):
     )
 
 
-class BaseLiquidHandlingRequest(BasePipettingRequest):
+class BaseLiquidHandlingData(BasePipettingData):
     """Base class for liquid handling requests."""
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/commands/pipetting_common.py]:
```suggestion
    """Base class for data payloads of commands that handle liquid."""
```

File: api/src/opentrons/protocol_engine/execution/pipetting.py [hunk 9]
```diff
@@ -33,8 +37,7 @@ async def pick_up_tip(
         """Pick up a tip at the specified "well"."""
         # get mount and config data from state and hardware controller
         hw_pipette = self._state.pipettes.get_hardware_pipette(
-            pipette_id=pipette_id,
-            attached_pipettes=self._hardware.attached_instruments
+            pipette_id=pipette_id, attached_pipettes=self._hardware.attached_instruments
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/execution/pipetting.py]:
```suggestion
            pipette_id=pipette_id,
            attached_pipettes=self._hardware.attached_instruments,
```

File: api/src/opentrons/protocol_engine/execution/pipetting.py [hunk 10]
```diff
@@ -57,17 +60,15 @@ async def pick_up_tip(
...
             presses=None,
-            increment=None
+            increment=None,
         )
 
         # after a successful pickup, update the hardware controller state
         self._hardware.set_current_tiprack_diameter(
-            mount=hw_pipette.mount,
-            tiprack_diameter=tip_geometry.diameter
+            mount=hw_pipette.mount, tiprack_diameter=tip_geometry.diameter
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/execution/pipetting.py]:
```suggestion
            mount=hw_pipette.mount,
            tiprack_diameter=tip_geometry.diameter,
```

File: api/src/opentrons/protocol_engine/execution/pipetting.py [hunk 11]
```diff
@@ -57,17 +60,15 @@ async def pick_up_tip(
...
         )
         self._hardware.set_working_volume(
-            mount=hw_pipette.mount,
-            tip_volume=tip_geometry.volume
+            mount=hw_pipette.mount, tip_volume=tip_geometry.volume
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/execution/pipetting.py]:
```suggestion
            mount=hw_pipette.mount,
            tip_volume=tip_geometry.volume,
```

File: api/src/opentrons/protocol_engine/execution/pipetting.py [hunk 12]
```diff
@@ -79,8 +80,7 @@ async def drop_tip(
         """Drop a tip at the specified "well"."""
         # get mount and config data from state and hardware controller
         hw_pipette = self._state.pipettes.get_hardware_pipette(
-            pipette_id=pipette_id,
-            attached_pipettes=self._hardware.attached_instruments
+            pipette_id=pipette_id, attached_pipettes=self._hardware.attached_instruments
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/execution/pipetting.py]:
```suggestion
            pipette_id=pipette_id,
            attached_pipettes=self._hardware.attached_instruments,
```

File: api/src/opentrons/protocol_engine/execution/pipetting.py [hunk 13]
```diff
@@ -115,8 +115,7 @@ async def aspirate(
         """Aspirate liquid from a well."""
         # get mount and config data from state and hardware controller
         hw_pipette = self._state.pipettes.get_hardware_pipette(
-            pipette_id=pipette_id,
-            attached_pipettes=self._hardware.attached_instruments
+            pipette_id=pipette_id, attached_pipettes=self._hardware.attached_instruments
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/execution/pipetting.py]:
```suggestion
            pipette_id=pipette_id,
            attached_pipettes=self._hardware.attached_instruments,
```

File: api/src/opentrons/protocol_engine/execution/pipetting.py [hunk 14]
```diff
@@ -139,9 +138,8 @@ async def aspirate(
             # set our current deck location to the well now that we've made
             # an intermediate move for the "prepare for aspirate" step
             current_location = DeckLocation(
-                pipette_id=pipette_id,
-                labware_id=labware_id,
-                well_name=well_name)
+                pipette_id=pipette_id, labware_id=labware_id, well_name=well_name
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/execution/pipetting.py]:
```suggestion
                pipette_id=pipette_id,
                labware_id=labware_id,
                well_name=well_name,
```

File: api/src/opentrons/protocol_engine/execution/pipetting.py [hunk 15]
```diff
@@ -165,8 +163,7 @@ async def dispense(
     ) -> float:
         """Dispense liquid to a well."""
         hw_pipette = self._state.pipettes.get_hardware_pipette(
-            pipette_id=pipette_id,
-            attached_pipettes=self._hardware.attached_instruments
+            pipette_id=pipette_id, attached_pipettes=self._hardware.attached_instruments
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/execution/pipetting.py]:
```suggestion
            pipette_id=pipette_id,
            attached_pipettes=self._hardware.attached_instruments,
```

File: api/src/opentrons/protocol_engine/commands/add_labware_definition.py [hunk 16]
```diff
@@ -37,18 +34,39 @@ class AddLabwareDefinitionResult(BaseModel):
...
+):
+    """Add labware definition command resource."""
+
+    commandType: AddLabwareDefinitionCommandType = "addLabwareDefinition"
+    data: AddLabwareDefinitionData
+    result: Optional[AddLabwareDefinitionResult]
+
+    _ImplementationCls: Type[
+        AddLabwareDefinitionImplementation
+    ] = AddLabwareDefinitionImplementation
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/commands/add_labware_definition.py:63]: If we wanted to avoid a magically-named `_ImplementationCls` attribute, we could do something like this:

```python
@command_mapper.map_to_implementation(AddLabwareDefinitionImplementation)
class AddLabwareDefinition(...)
    # ...
    # _ImplementationCls automatically added by class decorator.
```

This would make the link to the `command_mapper` module explicit.

I'm totally fine with it staying as-is for the sake of simplicity, though.

File: api/src/opentrons/protocol_engine/commands/command_mapper.py [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+
+    @staticmethod
+    def map_request_to_command(
+        request: CommandRequest,
+        command_id: str,
+        created_at: datetime,
+    ) -> Command:
+        """Map a CommandRequest instance to a full command."""
+        # NOTE(mc, 2021-06-22): mypy has trouble with this automatic
+        # request > command routing, but behavior is covered by unit tests
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/commands/command_mapper.py]: Obnoxious nitpick, sorry:

```suggestion
        # mypy has trouble with this automatic
        # request > command mapping, but behavior is covered by unit tests
```

* For comments like this, where they primarily explain or justify the code, I tend to omit the `NOTE(...)` signature. The author and timestamp aren't as important as they are in `TODO`s. For example, it's not like there are architectural details in your head about how to avoid this particular thing, and it's not like there's a specific MyPy update at the time of writing that we're waiting for to fix this. (I assume.)
* `routing` -> `mapping` to consistently use one verb for what this method does.

File: api/src/opentrons/protocol_engine/commands/command_mapper.py [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+            id=command_id,
+            createdAt=created_at,
+            status=CommandStatus.QUEUED,
+            data=request.data,  # type: ignore[arg-type]
+        )
+
+    @staticmethod
+    def update_command(command: Command, **update: Any) -> Command:
+        """Map a Command to a new Command instance with a different status."""
+        return command.copy(update=update)
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/commands/command_mapper.py]: I'm a little confused by this. AFAICT the status, specifically, isn't special here, as the docstring implies, right?

File: api/src/opentrons/protocol_engine/clients/transports.py [hunk 19]
```diff
@@ -65,16 +66,21 @@ def state(self) -> StateView:
...
-        if isinstance(command_state, FailedCommand):
-            raise command_state.error
+        if command.error is not None:
+            # TODO(mc, 2021-06-21): refactor when command.error is error details
+            # rather than a string
+            raise ProtocolEngineError(command.error)
 
-        return command_state.result
+        if command.result is None:
+            raise ValueError(f"Missing result in {command}; this is a bug")
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/clients/transports.py]: Since there's no legitimate reason for calling code to catch this error, should this instead be `assert command.result is not None`?

File: api/src/opentrons/protocol_engine/commands/command.py [hunk 20]
```diff
@@ -1,100 +1,92 @@
...
-    completed_at: datetime
-    request: ReqT
-    result: ResT
+class CommandStatus(str, Enum):
+    """Command execution status."""
 
+    QUEUED = "queued"
+    RUNNING = "running"
+    EXECUTED = "executed"
+    FAILED = "failed"
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/commands/command.py]: 1. Big fan of calling this `QUEUED` instead of `PENDING`.
2. Instead of `EXECUTED`/`FAILED`, maybe `SUCCEEDED`/`FAILED`? Both for symmetry, and because a failed command did, in a sense, execute?

File: api/src/opentrons/protocol_engine/commands/command.py [hunk 21]
```diff
@@ -1,100 +1,92 @@
...
+        None,
+        description="Command execution start timestamp, if started",
+    )
+    completedAt: Optional[datetime] = Field(
+        None,
+        description="Command execution completed timestamp, if completed",
+    )
+
+
+CommandT = TypeVar("CommandT", bound=BaseCommand)
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/commands/command.py]: No preference from me, bu is this intentionally separated from `CommandDataT` and `CommandResultT` above?

File: api/src/opentrons/protocol_engine/commands/command.py [hunk 22]
```diff
@@ -1,100 +1,92 @@
...
+            "Specific command type that determines data requirements and "
+            "execution behavior"
+        ),
+    )
+    status: CommandStatus = Field(..., description="Command execution status")
+    data: CommandDataT = Field(..., description="Command execution data payload")
+    result: Optional[CommandResultT] = Field(
+        None,
+        description="Command execution result data, if completed",
+    )
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/commands/command.py]:
```suggestion
    result: Optional[CommandResultT] = Field(
        None,
        description="Command execution result data, if succeeded",
    )
```
Very nitty disambiguation nitpick. (Did a "failed" command "complete"?)

File: api/src/opentrons/protocol_engine/commands/command.py [hunk 23]
```diff
@@ -1,100 +1,92 @@
...
+        description="Command execution failure, if failed",
+    )
+    startedAt: Optional[datetime] = Field(
+        None,
+        description="Command execution start timestamp, if started",
+    )
+    completedAt: Optional[datetime] = Field(
+        None,
+        description="Command execution completed timestamp, if completed",
+    )
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/commands/command.py:79]: Is this field also now doing what `failedAt` used to do?

File: api/src/opentrons/protocol_engine/__init__.py [hunk 24]
```diff
@@ -19,16 +19,15 @@
 )
 
 __all__ = [
-    # main export
+    # main factory and interface exports
+    "create_protocol_engine",
     "ProtocolEngine",
+    # error types
+    "ProtocolEngineError",
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/__init__.py:26]: I don't have an informed preference, but does this deliberately exclude more specific errors like `CommandDoesNotExistError`?

**Review by SyntaxColoring — COMMENTED** [review]
Partial review. What I've read so far looks good! Just some small questions and comments. I'm up to `api/src/opentrons/protocol_engine/execution/command_executor.py`.

I only skimmed the `drop_tip.py`, `load_labware.py`, etc. files because their changes seemed rote.

File: api/src/opentrons/protocol_engine/commands/add_labware_definition.py (hunk 16 above)
Comment (mcous) [reply api/src/opentrons/protocol_engine/commands/add_labware_definition.py:63]: I think that's a good idea to keep in the back pocket. For now, something like this would introduce a circular dependency between `protocol_engine.commands` and `protocol_engine.execution` that would need untangling.

What I like about this, though, is that I think the `AbstractCommandImplementation` base class should be coming in from `protocol_engine.execution`, rather than its current home of `protocol_engine.commands`. So I think it's worthwhile to untangle this and figure out a more explicit connection

File: api/src/opentrons/protocol_engine/commands/command.py (hunk 20 above)
Comment (mcous) [reply api/src/opentrons/protocol_engine/commands/command.py]: I grabbed these from existing sessions: https://github.com/Opentrons/opentrons/blob/fe4d6db248b2444506e839005e54bf4475d1bdc8/robot-server/robot_server/service/session/models/command.py#L51-L55

I like `SUCCEEDED` better, too, so gonna check real quick if we actually use `executed` anywhere at all, and make the change if we do not use it.

File: api/src/opentrons/protocol_engine/commands/command.py (hunk 23 above)
Comment (mcous) [reply api/src/opentrons/protocol_engine/commands/command.py:79]: Yup. "Completed" feels like it covers both success and failure, so I didn't feel like another field for `failedAt` was super necessary

File: api/src/opentrons/protocol_engine/__init__.py (hunk 24 above)
Comment (mcous) [reply api/src/opentrons/protocol_engine/__init__.py:26]: Not super deliberately, but the current story around PE errors needs some work. I could see all specific errors moving up to this level

File: api/src/opentrons/protocol_engine/commands/command_mapper.py (hunk 17 above)
Comment (mcous) [reply api/src/opentrons/protocol_engine/commands/command_mapper.py]: I think my intention here was to be more of a `TODO`; it feels conceivable that we could figure out some way to precisely type this (or something about a future mypy update makes it possible). I'll shift it to a TODO and make the `s/routing/mapping/` change

File: api/src/opentrons/protocol_engine/commands/command.py (hunk 20 above)
Comment (sanni-t) [reply api/src/opentrons/protocol_engine/commands/command.py]: +1 for `SUCCEEDED`

File: api/src/opentrons/protocol_engine/state/commands.py [hunk 25]
```diff
@@ -40,12 +41,15 @@ def __init__(self, state: CommandState) -> None:
...
 
-    def get_command_by_id(self, uid: str) -> Optional[CommandType]:
+    def get_command_by_id(self, command_id: str) -> Command:
         """Get a command by its unique identifier."""
         # TODO(mc, 2021-06-17): raise on missing ID, to line up with other state views
-        return self._state.commands_by_id.get(uid)
+        try:
+            return self._state.commands_by_id[command_id]
+        except KeyError:
+            raise CommandDoesNotExistError(f"Command {command_id} does not exist")
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/state/commands.py]: I think you can get rid of this `TODO` now. 🎉

File: api/src/opentrons/protocol_engine/state/geometry.py [hunk 26]
```diff
@@ -173,8 +172,13 @@ def get_tip_drop_location(
         pipette_config: PipetteDict,
     ) -> WellLocation:
         """Get tip drop location given labware and hardware pipette."""
-        # return to top if labware is fixed trash
-        if labware_id == FIXED_TRASH_ID:
+        # return to top if labware is fixed trash=
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/state/geometry.py]:
```suggestion
        # return to top if labware is fixed trash
```

Typo

File: api/src/opentrons/protocol_engine/state/commands.py [hunk 27]
```diff
@@ -40,12 +41,15 @@ def __init__(self, state: CommandState) -> None:
         """Initialize the view of command state with its underlying data."""
         self._state = state
 
-    def get_command_by_id(self, uid: str) -> Optional[CommandType]:
+    def get_command_by_id(self, command_id: str) -> Command:
         """Get a command by its unique identifier."""
         # TODO(mc, 2021-06-17): raise on missing ID, to line up with other state views
```
Comment (sanni-t) [inline api/src/opentrons/protocol_engine/state/commands.py]: Remove TODO?

File: api/src/opentrons/protocol_engine/state/labware.py [hunk 28]
```diff
@@ -77,7 +73,7 @@ def __init__(
             deck_definition=deck_definition,
         )
 
-    def handle_completed_command(self, command: CompletedCommandType) -> None:
+    def handle_command(self, command: Command) -> None:
         """Modify state in reaction to a completed command."""
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/state/labware.py]: Since we no longer have the type system to protect us, in functions like this that are meant to be run only on commands with a specific state, should we `assert command.status == SUCCEEDED` or something?

(Same goes for `PipetteState.handle_command()`, etc.)

Comment (SyntaxColoring) [reply api/src/opentrons/protocol_engine/state/labware.py]: Oh wait I just saw `state.py`. Is this docstring just out-of-date now—this is called for all commands, not just completed ones?

File: api/src/opentrons/protocols/runner/json_proto/command_translator.py [hunk 29]
```diff
@@ -1,50 +1,56 @@
...
+        self,
+        protocol: models.JsonProtocol,
+    ) -> List[pe_commands.CommandRequest]:
         """Return all Protocol Engine commands required to run the given protocol."""
-        result: List[pe.commands.CommandRequestType] = []
+        result: List[pe_commands.CommandRequest] = []
 
         for pipette_id, pipette in protocol.pipettes.items():
-            result += [self._translate_load_pipette(pipette_id, pipette)]
+            result.append(self._translate_load_pipette(pipette_id, pipette))
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocols/runner/json_proto/command_translator.py:33]: For my future reference, was there something wrong with `+=`, or is this just stylistic / for clarity?

File: api/src/opentrons/protocols/runner/json_proto/command_translator.py [hunk 30]
```diff
@@ -1,50 +1,56 @@
...
-from opentrons import protocol_engine as pe
+from opentrons.types import DeckSlotName, MountType
 from opentrons.protocols import models
+from opentrons.protocol_engine import (
+    commands as pe_commands,
+    DeckSlotLocation,
+    PipetteName,
+    WellLocation,
+    WellOrigin,
+)
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocols/runner/json_proto/command_translator.py:11]: This is nice. 👍

**Review by SyntaxColoring — COMMENTED** [review]
Another partial review. Still looks good so far. I'm just hitting the tests now.

File: api/tests/opentrons/protocol_engine/commands/test_command_mapper.py [hunk 31]
```diff
@@ -0,0 +1,194 @@
...
+    PipetteName,
+    WellLocation,
+)
+
+
+@pytest.mark.parametrize(
+    ("request_data", "request_cls", "command_cls"),
+    [
+        (
+            commands.AddLabwareDefinitionData.construct(
```
Comment (sanni-t) [inline api/tests/opentrons/protocol_engine/commands/test_command_mapper.py:22]: Why `.construct`?

File: api/tests/opentrons/protocol_engine/commands/test_command_mapper.py [hunk 32]
```diff
@@ -0,0 +1,194 @@
...
+                definition=cast(LabwareDefinition, {"mockDefintion": True})
```
Comment (sanni-t) [inline api/tests/opentrons/protocol_engine/commands/test_command_mapper.py]: 🤔

File: api/tests/opentrons/protocol_engine/commands/test_command_mapper.py [hunk 33]
```diff
@@ -0,0 +1,194 @@
...
+        status=commands.CommandStatus.RUNNING,
+        data=data,
+    )
+
+    # check that returned command is a new instance and input was not mutated
+    assert result is not command
+
+
+def test_update_command_timestamps() -> None:
+    """It should update a command status."""
```
Comment (sanni-t) [inline api/tests/opentrons/protocol_engine/commands/test_command_mapper.py]:
```suggestion
    """It should update a command fields like status and timestamps."""
```

**Review by sanni-t — COMMENTED** [review]
continuing..

File: api/tests/opentrons/protocol_engine/commands/test_drop_tip.py [hunk 34]
```diff
@@ -1,49 +1,46 @@
...
-    assert result
+from opentrons.protocol_engine.commands.drop_tip import (
+    DropTipData,
+    DropTipResult,
+    DropTipImplementation,
+)
 
 
-async def test_pick_up_tip_implementation(mock_handlers: AsyncMock) -> None:
+async def test_pick_up_tip_implementation(
```
Comment (sanni-t) [inline api/tests/opentrons/protocol_engine/commands/test_drop_tip.py]:
```suggestion
async def test_drop_tip_implementation(
```

File: api/tests/opentrons/protocol_engine/commands/test_drop_tip.py [hunk 35]
```diff
@@ -1,49 +1,46 @@
...
+    decoy: Decoy,
+    equipment: EquipmentHandler,
+    movement: MovementHandler,
+    pipetting: PipettingHandler,
+) -> None:
     """A DropTipRequest should have an execution implementation."""
```
Comment (sanni-t) [inline api/tests/opentrons/protocol_engine/commands/test_drop_tip.py]:
```suggestion
    """A DropTip command should have an execution implementation."""
```

File: api/tests/opentrons/protocol_engine/commands/test_load_pipette.py [hunk 36]
```diff
@@ -1,51 +1,49 @@
...
-
-
-async def test_load_pipette_implementation(mock_handlers: AsyncMock) -> None:
+async def test_load_pipette_implementation(
+    decoy: Decoy,
+    equipment: EquipmentHandler,
+    movement: MovementHandler,
+    pipetting: PipettingHandler,
+) -> None:
     """A LoadPipetteRequest should have an execution implementation."""
```
Comment (sanni-t) [inline api/tests/opentrons/protocol_engine/commands/test_load_pipette.py]:
```suggestion
    """A LoadPipette command should have an execution implementation."""
```

File: api/tests/opentrons/protocol_engine/commands/test_move_to_well.py [hunk 37]
```diff
@@ -1,49 +1,45 @@
...
 
 
-async def test_move_to_well_implementation(mock_handlers: AsyncMock) -> None:
+async def test_move_to_well_implementation(
+    decoy: Decoy,
+    equipment: EquipmentHandler,
+    movement: MovementHandler,
+    pipetting: PipettingHandler,
+) -> None:
     """A MoveToWellRequest should have an execution implementation."""
```
Comment (sanni-t) [inline api/tests/opentrons/protocol_engine/commands/test_move_to_well.py]:
```suggestion
    """A MoveToWell command should have an execution implementation."""
```

File: api/tests/opentrons/protocol_engine/commands/test_pick_up_tip.py [hunk 38]
```diff
@@ -1,49 +1,46 @@
...
 
 
-async def test_pick_up_tip_implementation(mock_handlers: AsyncMock) -> None:
+async def test_pick_up_tip_implementation(
+    decoy: Decoy,
+    equipment: EquipmentHandler,
+    movement: MovementHandler,
+    pipetting: PipettingHandler,
+) -> None:
     """A PickUpTipRequest should have an execution implementation."""
```
Comment (sanni-t) [inline api/tests/opentrons/protocol_engine/commands/test_pick_up_tip.py]:
```suggestion
    """A PickUpTip command should have an execution implementation."""
```

File: api/tests/opentrons/protocol_engine/commands/test_command_mapper.py (hunk 32 above)
Comment (mcous) [reply api/tests/opentrons/protocol_engine/commands/test_command_mapper.py]:
```suggestion
                # TODO(mc, 2021-06-25): do not mock out LabwareDefinition
                definition=cast(LabwareDefinition, {"mockDefinition": True})
```

File: api/tests/opentrons/protocol_engine/commands/test_command_mapper.py (hunk 31 above)
Comment (mcous) [reply api/tests/opentrons/protocol_engine/commands/test_command_mapper.py:22]: As discussed, to bypass validation and allow the mock `LabwareDefinition` below to be used. Added a TODO to remove the mock value

**Review by SyntaxColoring — APPROVED** [review]
Remainder of review. LGTM 👍 , modulo the above comments.

File: api/src/opentrons/protocol_engine/state/labware.py (hunk 28 above)
Comment (mcous) [reply api/src/opentrons/protocol_engine/state/labware.py]: Oops, yeah, out of date docstring

```suggestion
        """Modify state in reaction to a command."""
```

File: api/src/opentrons/protocols/runner/json_proto/command_translator.py (hunk 29 above)
Comment (mcous) [reply api/src/opentrons/protocols/runner/json_proto/command_translator.py:33]: Ah yeah, thoughts:

- I like the intent of `append` a little more, the wrapping `[ ... ]` on the value being appended bothers me
- `result = list_a + list_b` is going to create two new lists for every mutation to the one list we actually care about: `result`
    - It's a bit of a pre-optimization, but combined with the readability, when we're building up a new collection internally, I think it's usually more straightforward to mutate our `result` list

File: api/tests/opentrons/protocol_engine/conftest.py [hunk 39]
```diff
@@ -1,117 +1,11 @@
...
-
-
-@pytest.fixture
-def even_later(later: datetime) -> datetime:
-    """Get a future time."""
-    return later + timedelta(minutes=42)
-
-
-@pytest.fixture
-def mock_state_store() -> MagicMock:
```
Comment (sanni-t) [inline api/tests/opentrons/protocol_engine/conftest.py:61]: What was the reason for moving some of these frequently used fixtures out of conftest and adding them to individual test files?

**Review by sanni-t — APPROVED** [review]
Just the one question.
🎮

File: api/tests/opentrons/protocol_engine/conftest.py (hunk 39 above)
Comment (mcous) [reply api/tests/opentrons/protocol_engine/conftest.py:61]: This particular (MagicMock) fixture was only used in one test, so I moved it there.

In general, though, I'm finding conftest fixtures kind of hard to follow a lot of the time, and some tests become a lot more readable even if the fixtures are duplicated

Comment (SyntaxColoring) [reply api/tests/opentrons/protocol_engine/conftest.py:61]: Yeah, `conftest` feels too implicit a lot of the time, for me. Especially when there are multiple levels of `conftest`s.

I often think about turning fixtures into just regular functions that we `import` from some `test_utils` module. That would solve the implicitness problem.
